### PR TITLE
Add clear browsing data logic

### PR DESCRIPTION
**Pull request by codeimpl**
User requests to clear browsing data when exit browser. This patch 
implements the logic of the feature.

Refs: #768

### CODE REVIEW

File: banana_cake/java/org/banana/cake/CakeAutoClearBrowsingData.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package org.banana.cake;
+
+import org.banana.cake.interfaces.BananaAutoClearBorwsingData;
+
+import org.chromium.chrome.browser.browsing_data.BrowsingDataBridge;
+
+class CakeCommandLine implements BananaAutoClearBorwsingData {
```
Comment (romandev) [inline banana_cake/java/org/banana/cake/CakeAutoClearBrowsingData.java:12]: CakeAutoClearBrowsingData

File: banana_cake/interfaces/java/org/banana/cake/interfaces/BananaAutoClearBrowsingData.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+import android.view.Menu;
+import android.widget.FrameLayout;
+
+import org.chromium.chrome.browser.browsing_data.BrowsingDataType;
+import org.chromium.chrome.browser.browsing_data.TimePeriod;
+
+public interface BananaAutoClearBrowsingData {
+    public interface BananaBrowsingDataType extends BrowsingDataType {}
+    public interface BananaTimePeriod extends TimePeriod {}
```
Comment (romandev) [inline banana_cake/interfaces/java/org/banana/cake/interfaces/BananaAutoClearBrowsingData.java:16]: I'm not sure but there is not reason to expose these interfaces to upper layer?

File: banana_cake/interfaces/java/org/banana/cake/interfaces/BananaAutoClearBrowsingData.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+    static BananaAutoClearBrowsingData get() {
+        return BananaInterfaceProvider.get(BananaAutoClearBrowsingData.class);
+    }
+
+    void clearBrowsingDataIfNeeded();
+    void clearBrowsingData();
```
Comment (romandev) [inline banana_cake/interfaces/java/org/banana/cake/interfaces/BananaAutoClearBrowsingData.java:23]: What's difference between two methods? Can we have only one version?

File: banana_cake/interfaces/java/org/banana/cake/interfaces/BananaAutoClearBrowsingData.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package org.banana.cake.interfaces;
+
+import android.view.Menu;
+import android.widget.FrameLayout;
+
+import org.chromium.chrome.browser.browsing_data.BrowsingDataType;
+import org.chromium.chrome.browser.browsing_data.TimePeriod;
```
Comment (romandev) [inline banana_cake/interfaces/java/org/banana/cake/interfaces/BananaAutoClearBrowsingData.java:12]: Looks unnecessary imports?

File: src/org/triple/banana/browsing_data/AutoClearBrowsingData.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
+// Copyright 2019 The Triple Banana Authors. All rights reserved.
```
Comment (jdragonbae) [inline src/org/triple/banana/browsing_data/AutoClearBrowsingData.java:1]: // Copyright 2020

File: src/org/triple/banana/browsing_data/AutoClearBrowsingData.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+import org.banana.cake.interfaces.BananaAutoClearBrowsingData;
+import org.banana.cake.interfaces.BananaBrowsingDataBridge;
+import org.triple.banana.settings.ExtensionFeatures;
+import org.triple.banana.settings.ExtensionFeatures.FeatureName;
+
+public class AutoClearBrowsingData implements BananaAutoClearBrowsingData {
+    @Override
+    public void autoClearBrowsingData() {
+        if (ExtensionFeatures.isEnabled(FeatureName.AUTO_CLEAR_BROWSING_DATA, true)) {
```
Comment (jdragonbae) [inline src/org/triple/banana/browsing_data/AutoClearBrowsingData.java:16]: @codeimpl, @romandev  I wonder the default value of `AUTO_CLEAR_BROWSING_DATA` is true or false.

File: upstream/tb_chrome_android_java_src_org_chromium_chrome_browser_ChromeTabbedActivity.java.patch [hunk 7]
```diff
@@ -0,0 +1,12 @@
+diff --git tb/chrome/android/java/src/org/chromium/chrome/browser/ChromeTabbedActivity.java tb_diff/chrome/android/java/src/org/chromium/chrome/browser/ChromeTabbedActivity.java
+--- tb/chrome/android/java/src/org/chromium/chrome/browser/ChromeTabbedActivity.java
++++ tb_diff/chrome/android/java/src/org/chromium/chrome/browser/ChromeTabbedActivity.java
+@@ -1941,6 +1941,8 @@ public class ChromeTabbedActivity extends ChromeActivity<ChromeActivityComponent
+ 
+         final boolean shouldCloseTab = backShouldCloseTab(currentTab);
+ 
++        org.banana.cake.interfaces.BananaAutoClearBrowsingData.get().autoClearBrowsingData();
```
Comment (romandev) [inline upstream/tb_chrome_android_java_src_org_chromium_chrome_browser_ChromeTabbedActivity.java.patch:8]: We should call the `onExitBrowser` event here and then implement the clear data in the event.

File: banana_cake/interfaces/java/org/banana/cake/interfaces/BananaBrowsingDataBridge.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2020 The Triple Banana Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package org.banana.cake.interfaces;
+
+public interface BananaBrowsingDataBridge {
```
Comment (romandev) [inline banana_cake/interfaces/java/org/banana/cake/interfaces/BananaBrowsingDataBridge.java:8]: This should be `BananaClearBrowsingData`

File: banana_cake/interfaces/java/org/banana/cake/interfaces/BananaAutoClearBrowsingData.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2020 The Triple Banana Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package org.banana.cake.interfaces;
+
+public interface BananaAutoClearBrowsingData {
```
Comment (romandev) [inline banana_cake/interfaces/java/org/banana/cake/interfaces/BananaAutoClearBrowsingData.java:8]: It should have another name? BananaApplication... what? I'm not sure.

File: banana_cake/interfaces/java/org/banana/cake/interfaces/BananaAutoClearBrowsingData.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+    static BananaAutoClearBrowsingData get() {
+        return BananaInterfaceProvider.get(BananaAutoClearBrowsingData.class);
+    }
+
+    void autoClearBrowsingData();
```
Comment (romandev) [inline banana_cake/interfaces/java/org/banana/cake/interfaces/BananaAutoClearBrowsingData.java:13]: onExit()? onTerminate()? and so on?

File: src/org/triple/banana/browsing_data/AutoClearBrowsingData.java (hunk 6 above)
Comment (romandev) [reply src/org/triple/banana/browsing_data/AutoClearBrowsingData.java:16]: Default value should be false but why do we check isEnabled with default value here?

File: banana_cake/interfaces/java/org/banana/cake/interfaces/BananaBrowsingDataBridge.java (hunk 8 above)
Comment (codeimpl) [reply banana_cake/interfaces/java/org/banana/cake/interfaces/BananaBrowsingDataBridge.java:8]: done

File: banana_cake/interfaces/java/org/banana/cake/interfaces/BananaAutoClearBrowsingData.java (hunk 9 above)
Comment (codeimpl) [reply banana_cake/interfaces/java/org/banana/cake/interfaces/BananaAutoClearBrowsingData.java:8]: Changed it as BananaApplicationEvent.

File: banana_cake/interfaces/java/org/banana/cake/interfaces/BananaAutoClearBrowsingData.java (hunk 10 above)
Comment (codeimpl) [reply banana_cake/interfaces/java/org/banana/cake/interfaces/BananaAutoClearBrowsingData.java:13]: Changed it as onTerminate().

File: src/org/triple/banana/browsing_data/AutoClearBrowsingData.java (hunk 5 above)
Comment (codeimpl) [reply src/org/triple/banana/browsing_data/AutoClearBrowsingData.java:1]: done

File: src/org/triple/banana/browsing_data/AutoClearBrowsingData.java (hunk 6 above)
Comment (codeimpl) [reply src/org/triple/banana/browsing_data/AutoClearBrowsingData.java:16]: done

File: upstream/tb_chrome_android_java_src_org_chromium_chrome_browser_ChromeTabbedActivity.java.patch (hunk 7 above)
Comment (codeimpl) [reply upstream/tb_chrome_android_java_src_org_chromium_chrome_browser_ChromeTabbedActivity.java.patch:8]: Changed it as org.banana.cake.interfaces.BananaApplicationEvent.get().onTerminate();

**Review by romandev — APPROVED** [review]
LGTM

**Pull request merged**